### PR TITLE
Makefile: add a few missing dependencies

### PR DESCRIPTION
**Pull request by nojb**
In this PR:
- Add missing dependencies when producing `stdlib/.depend`; and
- Add missing dependency of `ocaml{c,opt}.opt` on `stdlib/libasmrun.$(A)`

### CODE REVIEW

File: stdlib/Makefile [hunk 1]
```diff
@@ -266,7 +266,7 @@ GNUISH_SED = \
 
 .INTERMEDIATE: .depend.tmp
 
-.depend.tmp:
+.depend.tmp: $(wildcard *.mli *.ml)
```
Comment (dra27) [inline stdlib/Makefile:269]: This doesn't work (as CI is demonstrating) as it introduces an actual dependency between `.depend` on the files. It would alas appear I didn't test `make alldepend` in enough detail apropos https://github.com/ocaml/ocaml/pull/11126#pullrequestreview-917494259.

The fix is to change L275 below to:

```make
depend: .depend.tmp .depend
```

That means that when `Makefile` is loaded "normally" (e.g. `make -C stdlib all`), then the dependency of `.depend` on `.depend.tmp` is ignored (as `.depend.tmp` is intermediate) but when `make -C stdlib depend` is called, `.depend.tmp` is always built (because it's a direct dependency) but is still deleted the build (because `.INTERMEDIATE` files always are).

It's become a bit convoluted, though, just to avoid the explicit call to `rm -f .depend.tmp`, @shindere?

**Review by dra27 — COMMENTED** [review]
I haven't checked the correctness of the change in the main `Makefile`, but there's a fault with the stdlib one. I've given the correction, but it's making me wonder if the use of `.INTERMEDIATE` there ought to be reverted.

For the dependency of the compilers on the runtime libraries - I haven't checked, but is this a regression from #11248? If it's _not_ a regression then it would be worth considering this separately, because this may be one of the instances where this dependency is intentionally not present - IIUC, with this change, when working on the runtime `make all` will now rebuild the entire system?

**Comment by shindere** [comment]
Many thanks for the work, @nojb!

I do not understand why `ocamlc.opt` has to depend on
`stdlib/libasmrun.$(A)` whereas this file does not actually appear on
the command-line?

I assume there must be kind of an indirect dependency and I wonder
whether there would be a way to express it in a way that better reflects
the real dependencies? And by the way shouldn't it be
`stdlib/libcamlrun.$(A)` rather than `stdlib/libasmrun.$(A)`?



The same question applies to the recipe that builds `ocamlopt.opt`.

Regarding the change on `stdlib/libcamlrun.$(A)`: the new way looks
indeed more straightforward to me but I am not sure the change is safe
given how the current build system works. Especially I am not sure this
dependency is enough for our current bootstrap procedure to work. Have
you checked it, @nojb?

Same question for `stdlib/libasmrun.$(A)`. I think @dra27 and I did
consider this change when we reviewed the merge of `runtime/Makefile` in
this file, but then decided not to do it, at least as long as all the
makefiles have not been merged. So, to be honest, if these two changes
are not strictly necessary to fix bugs, I'd tend to suggest to leave
them out for now, wait until all the sub-makefiles have been merged and
then, once we only have one Makefile to deal with, it should hopefully become
easier to see what can be simplified without breaking anything.

In other words: in my understanding your two changes to these last rules
remove some dependencies and this may work but it's not for sure a
conservative change, so I'd rather not touch this, at the moment.

I trust @Êra27's opinion on the change in `stdlib/Makefile`.

**Comment by dra27** [comment]
> I do not understand why `ocamlc.opt` has to depend on `stdlib/libasmrun.$(A)` whereas this file does not actually appear on the command-line?

`stdlib/libasmrun.$(A)` will be linked by ocamlopt to create `ocamlc.opt` - so it depends on it in the same way as it also depends on `stdlib/stdlib.cmxa`. Definitely `libasmrun` - it's creating the `.opt` versions here, so we're in native mode.

**Comment by nojb** [comment]
Note: I haven't had time to go over @dra27's review yet, I will try to do so shortly...

**Comment by shindere** [comment]
David Allsopp (2022/08/02 09:26 -0700):
> > I do not understand why `ocamlc.opt` has to depend on `stdlib/libasmrun.$(A)` whereas this file does not actually appear on the command-line?
>
> `stdlib/libasmrun.$(A)` will be linked by ocamlopt to create
> `ocamlc.opt` - so it depends on it in the same way as it also depends
> on `stdlib/stdlib.cmxa`. Definitely `libasmrun` - it's creating the
> `.opt` versions here, so we're in native mode.

Don't we have a better way to express these dependencies? e.g. having
ocamlc.opt depend on an other file that more explicitly depends on
libasmrun.a or something like that?

Independently: perhaps we could avoir the call to the `filter` function
by introducing varialbes for each kind of dependencies. Then we could
use all the variables as list of prerequisites and just those that are
relevant in the recpie, rather than de-constructing with filter
something that has jut been built.

One other possibility: would it perhaps be possible to make libasmrun.a
appear explicitly on the command-line, although it would actually not be
strictly necessary?

Just asking, I am okay with almost any decision, I think.

**Comment by dra27** [comment]
Any program compiled with `ocamlopt` does have a direct, real dependency on `libasmrun.a` (as in, the code in `libasmrun.a` will appear in that program), it's just that it's not part of the command line.

I agree we could avoid the `filter` by having a the .cmxa files and the BYTESTART listed explicitly in a variable and use that in both the dependencies and then the command itself. I agree that `$^`, `$(filter ..)` etc. all look neat up to the point you then have to keep tweaking it because of adding something like this!

**Comment by shindere** [comment]
David Allsopp (2022/08/04 02:47 -0700):
> Any program compiled with `ocamlopt` does have a direct, real
> dependency on `libasmrun.a` (as in, the code in `libasmrun.a` will
> appear in that program), it's just that it's not part of the command
> line.

I understand what you say, yes.

What I meant is that, indeed, to compile a program in native mode you do
need _one_ native runtime. For now, the fact that the runtime to use is
that given version of `libasmrun.a` seems to be hard-coded in the
compiler. Of course there are the `-with/without-runtime` and
`-runtime-variant` options but in my opinion they are a bit limited. I
could imagine a much more general option, a bit like `-use-runtime`
which would let the user be explicit about which runtime library to use.
Then we could use this option in the recipes, have the name of the
library in a variable and even make it possible for this variable to be
overriden, which would make the whole system less rigid. What do you
think?

**Comment by nojb** [comment]
Not very well thought out. Closing.
